uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature

```diff
@@ -1740,6 +1740,102 @@ Feature: GGMQ-1
+      | aws.greengrass.clientdevices.IPDetector  | LATEST                                  |
+      | <agent>                                  | classpath:/local-store/recipes/<recipe> |
+    And I create client device "clientDeviceTest"
+    When I associate "clientDeviceTest" with ggc
+    And I update my Greengrass deployment configuration, setting the component aws.greengrass.Nucleus configuration to:
+    """
+{
+    "MERGE":{
+        "mqtt":{
+            "keepAliveTimeoutMs":10000,
+            "pingTimeoutMs":3000
+        },
+        "logging": {
+            "fileSizeKB": 102400,
+            "totalLogsSizeKB": 1024000
+        }
+    }
+}
+    """
+    And I update my Greengrass deployment configuration, setting the component aws.greengrass.clientdevices.Auth configuration to:
+    """
+{
+    "MERGE":{
+        "deviceGroups":{
+            "formatVersion":"2021-03-05",
+            "definitions":{
+                "MyPermissiveDeviceGroup":{
+                    "selectionRule":"thingName: ${clientDeviceTest}",
+                    "policyName":"MyPermissivePolicy"
+                }
+            },
+            "policies":{
+                "MyPermissivePolicy":{
+                    "AllowAll":{
+                        "statementDescription":"Allow client devices to perform all actions.",
+                        "operations":[
+                            "*"
+                        ],
+                        "resources":[
+                            "*"
+                        ]
+                    }
+                }
+            }
+        }
+    }
+}
+    """
+    And I update my Greengrass deployment configuration, setting the component <agent> configuration to:
+    """
+{
+    "MERGE":{
+        "controlAddresses":"${mqttControlAddresses}",
+        "controlPort":"${mqttControlPort}"
+    }
+}
+    """
+    And I deploy the Greengrass deployment configuration
+    Then the Greengrass deployment is COMPLETED on the device after 5 minutes
+    And the aws.greengrass.clientdevices.mqtt.EMQX log on the device contains the line "is running now!." within 1 minutes
+    And the greengrass log on the device contains the line "New server certificate generated" within 1 minutes
+    # Here we ensure Nucleus has generated initial certificate of broker but it possible not all IP addresses are added to it
+    #  sometimes detecting IP and updating cert takes >40 seconds
+    And I wait 65 seconds
+
+    And I discover core device broker as "default_broker" from "clientDeviceTest" in OTF
+    And I connect device "clientDeviceTest" on <agent> to "default_broker" using mqtt "<mqtt-v>"
+
+    Then I retrieve the certificate of broker "default_broker" and store as "BROKER_CERTIFICATE_BEFORE_DISCONNECT"
+
+    When I set device mqtt connectivity to offline
+    And the greengrass log on the device contains the line "com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Failed to connect to AWS IoT Core" at least 3 times within 2 minutes
```

Comment:
why do you want this 3 times? Why isn't 1 enough?

Reply:
At first time 1 is enough. Usually Nucleus makes 3 attempts in the short period of time like:
```
2023-07-25T20:43:30.342Z [ERROR] (Thread-10) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Failed to connect to AWS IoT Core. {clientId=gg-5e37c73c40370bb47008-ggc-thing, error=socket operation timed out.}
2023-07-25T20:43:34.858Z [ERROR] (Thread-10) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Failed to connect to AWS IoT Core. {clientId=gg-5e37c73c40370bb47008-ggc-thing, error=socket operation timed out.}
2023-07-25T20:43:37.958Z [ERROR] (Thread-10) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Failed to connect to AWS IoT Core. {clientId=gg-5e37c73c40370bb47008-ggc-thing, error=socket operation timed out.}
```

Reply:
So if you only need 1, why does this step require 3?

Reply:
I assumed Nucleus made series of attempts before falling to disconnected state.

Reply:
No.

https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/3b555bb0321a78c4919ef59dc12ccaef7793b394/src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java#L147-L162

Reply:
Ok, will change to 1

Reply:
onDisconnection() log another message - "Connection interrupted".
Message "Failed to connect to AWS IoT Core" we are waiting for logged from onConnectionFailure() which called 3 times after onDisconnection() .

But "Connection interrupted" is appeared already in log file.

Reply:
Yes, "Connection interrupted" means that the device was disconnected. If you see this, and there is no "Connection resumed" then it is offline.

Reply:
Fixed in 224fb32b63284172f8b7ee71ec05037613d78680

Reply:
At the moment has no steps to check log file miss some string.